man/ecfg.1.ronn
# ecfg(1) -- manage application secrets via encrypted config

## SYNOPSIS

`ecfg` [`-k`|`--keydir` *dir*] `command` [`args`]

## DESCRIPTION

`ecfg` is a utility for managing a collection of secrets, typically to be
committed to source control. The secrets are encrypted using public key,
elliptic curve cryptography. Secrets are collected in a JSON, YAML, or TOML
file, in which all the string values are encrypted. Public keys are embedded in
the file, and the decrypter looks up the corresponding private key from its
local filesystem or process environment.

See ecfg(5) for more information on the `ecfg` file format, and read on for
a workflow example.

## COMMANDS

`ecfg help` [*command*]

:   Show (this) help for `ecfg` in general, or for a specific command

`ecfg encrypt` : ecfg-encrypt(1)

:   Encrypt an `ecfg` file (alias: `ecfg e`)

`ecfg decrypt` : ecfg-decrypt(1)

:   Decrypt an `ecfg` file (alias: `ecfg d`)

`ecfg keygen` : ecfg-keygen(1)

:   Generate an `ecfg` keypair (alias: `ecfg g`)

## GLOBAL OPTIONS

`-k`, `--keydir`=*<dir>*

:   Use the provided directory instead of the default key paths (decribed in
    the KEY MANAGEMENT section)

## ENVIRONMENT

`ECFG_KEYDIR`

:   Use a custom directory instead of the default key lookup path decribed in
    the KEY MANAGEMENT section.

`ECFG_PRIVATE_KEY`

:   When decrypting, instead of looking up the matching private key for the
    public key given in the input file, assume the file was encrypted to the
    provided private key. This option is useful when running in environments
    such as heroku where obtaining keys from disk is impractical.

## KEY MANAGEMENT

`ecfg` keypairs are stored as individual files in a key directory. The file
name is the public key and the file content is the private key. `ecfg` has a
default lookup path for key directories:

* `$XDG_CONFIG_HOME/ecfg/keys` (if `$XDG_CONFIG_HOME` is set and running as non-root user)
* `$HOME/.ecfg/keys` (if running as non-root user)
* `/etc/ecfg/keys`
* `/opt/ejson/keys` (for backwards-compatibility with `ejson`)

When passing `-k` or `--keydir` to `ecfg`, or when invoked with `ECFG_KEYDIR`
in the environment, this lookup path is completely ignored and the key is
instead retrieved from or stored to the provided path.

If `ECFG_PRIVATE_KEY` is set for decryption, the key directories aren't even
touched; instead, we just assume the provided private key is the correct one,
failing if it's not.

## WORKFLOW

### 1: Create the Keydir

By default, `ecfg` looks for keys in `/opt/ecfg/keys`. You can change this by
setting `ECFG_KEYDIR` or passing the `-keydir` option.

```
$ mkdir -p /opt/ecfg/keys
```

### 2: Generate a keypair

When called with `-w`, `ecfg keygen` will write the keypair into the `keydir`
and print the public key. Without `-w`, it will print both keys to stdout. This
is useful if you have to distribute the key to multiple servers via
configuration management, etc.

```
$ ecfg keygen
Public Key:
63ccf05a9492e68e12eeb1c705888aebdcc0080af7e594fc402beb24cce9d14f
Private Key:
75b80b4a693156eb435f4ed2fe397e583f461f09fd99ec2bd1bdef0a56cf6e64

$ ./ecfg keygen -w
53393332c6c7c474af603c078f5696c8fe16677a09a711bba299a6c1c1676a59
$ cat /opt/ecfg/keys/5339*
888a4291bef9135729357b8c70e5a62b0bbe104a679d829cdbe56d46a4481aaf
```

### 3: Create an ecfg file

The format is described in more detail in ecfg(5). For now, create a file that
looks something like this. Fill in the `<key>` with whatever you got back in
step 2.

Create this file as `test.ecfg.json`:

```
{
  "_public_key": "<key>",
  "database_password": "1234password"
}
```

You can also use YAML or TOML if you'd prefer, as long as there's a
`_public_key` element at the top-level.

### 4: Encrypt the file

Running `ecfg encrypt test.ecfg.json` will encrypt any new plaintext keys in
the file, and leave any existing encrypted keys untouched:

```
{
  "_public_key": "63ccf05a9492e68e12eeb1c705888aebdcc0080af7e594fc402beb24cce9d14f",
  "database_password": "EJ[1:WGj2t4znULHT1IRveMEdvvNXqZzNBNMsJ5iZVy6Dvxs=:kA6ekF8ViYR5ZLeSmMXWsdLfWr7wn9qS:fcHQtdt6nqcNOXa97/M278RX6w==]"
}
```

Try adding another plaintext secret to the file and run `ecfg encrypt
test.ecfg.json` again. The `database_password` field will not be changed, but the
new secret will be encrypted.

### 5: Decrypt the file

To decrypt the file, you must have a file present in the `keydir` whose name is
the 64-byte hex-encoded public key exactly as embedded in the ecfg(5) document.
The contents of that file must be the similarly-encoded private key. If you used
`ecfg keygen -w`, you've already got this covered.

Unlike ecfg-encrypt(1), which overwrites the specified files, ecfg-decrypt(1)
only takes one file parameter, and prints the output to `stdout`:

```
$ ecfg decrypt foo.ecfg.json
{
  "_public_key": "63ccf05a9492e68e12eeb1c705888aebdcc0080af7e594fc402beb24cce9d14f",
  "database_password": "1234password"
}
```
TODO

## BUGS

Report security issues to <burke.libbey@shopify.com> and
<security@shopify.com>.

File non-security-related bugs at <https://github.com/Shopify/ecfg>.

## COPYRIGHT

`ecfg` is copyright (C) 2016 Shopify under MIT license.

## SEE ALSO

ecfg-encrypt(1), ecfg-decrypt(1), ecfg-keygen(1), ecfg(5)


CHANGELOG.md
# 0.3.0

* Add key lookup path (~/.ecfg/keys -> /etc/ecfg/keys -> /opt/ejson/keys)

# 0.2.0

* Add TOML support

# 0.1.0

* Initial release


ecfg.go
// Package ecfg implements the primary interface to interact with ecfg
// documents and keypairs. The CLI implemented by cmd/ecfg is a fairly thin
// wrapper around this package.
package ecfg

import (
	"encoding/hex"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"

	"github.com/Shopify/ecfg/pkg/crypto"
	"github.com/Shopify/ecfg/pkg/format"
	"github.com/Shopify/ecfg/pkg/json"
	"github.com/Shopify/ecfg/pkg/toml"
	"github.com/Shopify/ecfg/pkg/yaml"
)

type FileType int

const (
	FileTypeJSON = iota
	FileTypeYAML
	FileTypeTOML
)

// GenerateKeypair is used to create a new ecfg keypair. It returns the keys as
// hex-encoded strings, suitable for printing to the screen. hex.DecodeString
// can be used to load the true representation if necessary.
func GenerateKeypair() (pub string, priv string, err error) {
	var kp crypto.Keypair
	if err := kp.Generate(); err != nil {
		return "", "", err
	}
	return kp.PublicString(), kp.PrivateString(), nil
}

// EncryptFileInPlace takes a path to a file on disk, which must be a valid ecfg file
// (see README.md for more on what constitutes a valid ecfg file). Any
// encryptable-but-unencrypted fields in the file will be encrypted using the
// public key embdded in the file, and the resulting text will be written over
// the file present on disk.
func EncryptFileInPlace(filePath string, fileType FileType) (int, error) {
	data, err := readFile(filePath)
	if err != nil {
		return -1, err
	}

	fileMode, err := getMode(filePath)
	if err != nil {
		return -1, err
	}

	newdata, err := EncryptData(data, fileType)
	if err != nil {
		return -1, err
	}

	if err := writeFile(filePath, newdata, fileMode); err != nil {
		return -1, err
	}

	return len(newdata), nil
}

func EncryptData(data []byte, fileType FileType) ([]byte, error) {
	fh := handlerForType(fileType)

	var myKP crypto.Keypair
	if err := myKP.Generate(); err != nil {
		return nil, err
	}

	pubkey, err := fh.ExtractPublicKey(data)
	if err != nil {
		return nil, err
	}

	encrypter := myKP.Encrypter(pubkey)

	return fh.TransformScalarValues(data, encrypter.Encrypt)
}

// DecryptFile takes a path to an encrypted ecfg file and returns the data
// decrypted. The public key used to encrypt the values is embedded in the
// referenced document, and the matching private key is searched for in
// keypath. There must exist a file in at least one of the keypath entries
// whose name is the public key from the ecfg document, and whose contents are
// the corresponding private key. See README.md for more details on this.
func DecryptFile(filePath string, keypath []string, fileType FileType) ([]byte, error) {
	data, err := readFile(filePath)
	if err != nil {
		return nil, err
	}

	return DecryptData(data, keypath, fileType)
}

// DecryptData takes a an encrypted ecfg document and returns the same
// document, decrypted. The public key used to encrypt the values is embedded
// in the document, and the matching private key is searched for in keypath.
// There must exist a file in at least one of the keypath entries whose name is
// the public key from the ecfg document, and whose contents are the
// corresponding private key. See README.md for more details on this.
func DecryptData(data []byte, keypath []string, fileType FileType) ([]byte, error) {
	fh := handlerForType(fileType)

	pubkey, err := fh.ExtractPublicKey(data)
	if err != nil {
		return nil, err
	}

	privkey, err := findPrivateKey(pubkey, keypath)
	if err != nil {
		return nil, err
	}

	myKP := crypto.Keypair{
		Public:  pubkey,
		Private: privkey,
	}

	decrypter := myKP.Decrypter()

	return fh.TransformScalarValues(data, decrypter.Decrypt)
}

// DefaultKeypath is UserKeypath prefixed to SystemKeypath. For root, this will
// be equal to SystemKeypath, and for other users, this will cause key lookups
// to first try their own local keys, falling back to system keys if that
// fails.
func DefaultKeypath() (keypath []string) {
	for _, elem := range UserKeypath() {
		keypath = append(keypath, elem)
	}
	for _, elem := range SystemKeypath() {
		keypath = append(keypath, elem)
	}
	return
}

// UserKeypath returns the user-specific locations at which to search for ecfg
// keys. In most cases, this is empty for root, and ~/.ecfg/keys in other cases.
// If XDG_CONFIG_HOME is set, $XDG_CONFIG_HOME/ecfg/keys is highest priority.
func UserKeypath() (keypath []string) {
	// no user keypath entries for root.
	if getuid() == 0 {
		return
	}

	xdgConfigHome := getenv("XDG_CONFIG_HOME")
	if xdgConfigHome != "" {
		keypath = append(keypath, filepath.Join(xdgConfigHome, "ecfg", "keys"))
	}
	keypath = append(keypath, filepath.Join(getenv("HOME"), ".ecfg", "keys"))
	return
}

// SystemKeypath returns the default system-wide locations at which to search
// for ecfg keys. /opt/ejson/keys is provided for backwards-compatibility with
// ejson.
func SystemKeypath() (keypath []string) {
	keypath = append(keypath, "/etc/ecfg/keys")
	keypath = append(keypath, "/opt/ejson/keys")
	return
}

func findPrivateKey(pubkey [32]byte, keypath []string) (privkey [32]byte, err error) {
	keyString := os.Getenv("ECFG_PRIVATE_KEY")
	if keyString == "" {
		for _, keydir := range keypath {
			keyFile := fmt.Sprintf("%s/%x", keydir, pubkey)
			fileContents, err := readFile(keyFile)
			if err == nil {
				keyString = strings.TrimSpace(string(fileContents))
				break
			}
		}
	}
	if keyString == "" {
		err = fmt.Errorf("private key not found in keypath")
		return
	}

	bs, err := hex.DecodeString(keyString)
	if err != nil {
		return
	}

	if len(bs) != 32 {
		err = fmt.Errorf("invalid private key retrieved from keydir")
		return
	}

	copy(privkey[:], bs)
	return
}

func handlerForType(typ FileType) format.FormatHandler {
	switch typ {
	case FileTypeJSON:
		return &json.FormatHandler{}
	case FileTypeYAML:
		return &yaml.FormatHandler{}
	case FileTypeTOML:
		return &toml.FormatHandler{}
	default:
		panic("bug: invalid file type")
	}
}

// for mocking in tests
func _getMode(path string) (os.FileMode, error) {
	fi, err := os.Stat(path)
	if err != nil {
		return 0, err
	}
	return fi.Mode(), nil
}

// for mocking in tests
var (
	readFile  = ioutil.ReadFile
	writeFile = ioutil.WriteFile
	getMode   = _getMode
	getuid    = os.Getuid
	getenv    = os.Getenv
)
